hummingbot/connector/derivative/phemex_perpetual/phemex_perpetual_api_order_book_data_source.py

```diff
@@ -0,0 +1,233 @@
+        elif self._trade_messages_queue_key in event_message:
+            channel = self._trade_messages_queue_key
+        elif event_message.get("method", None) == self._funding_info_messages_queue_key:
+            channel = self._funding_info_messages_queue_key
+        return channel
+
+    async def _parse_order_book_diff_message(self, raw_message: Dict[str, Any], message_queue: asyncio.Queue):
+        timestamp: float = time.time()
+        raw_message["symbol"] = await self._connector.trading_pair_associated_to_exchange_symbol(raw_message["symbol"])
+        if raw_message["type"] == "incremental":
+            book = raw_message[self._diff_messages_queue_key]
+            order_book_message: OrderBookMessage = OrderBookMessage(
+                OrderBookMessageType.DIFF,
+                {
+                    "trading_pair": raw_message["symbol"],
+                    "update_id": raw_message["sequence"],
+                    "bids": book["bids"],
+                    "asks": book["asks"],
+                },
+                timestamp=timestamp,
+            )
+            message_queue.put_nowait(order_book_message)
+
+    async def _parse_trade_message(self, raw_message: Dict[str, Any], message_queue: asyncio.Queue):
+        raw_message["symbol"] = await self._connector.trading_pair_associated_to_exchange_symbol(raw_message["symbol"])
+        for trade in raw_message[self._trade_messages_queue_key]:
+            mapped_trade = TradeStructure(*trade)
+            trade_message: OrderBookMessage = OrderBookMessage(
+                OrderBookMessageType.TRADE,
+                {
+                    "trading_pair": raw_message["symbol"],
+                    "trade_type": mapped_trade.side.upper(),
+                    "trade_id": mapped_trade.timestamp,
+                    "update_id": raw_message["sequence"],
+                    "price": mapped_trade.price,
+                    "amount": mapped_trade.amount,
+                },
+                timestamp=mapped_trade.timestamp // 1e9,
+            )
+
+            message_queue.put_nowait(trade_message)
+
+    async def listen_for_order_book_snapshots(self, ev_loop: asyncio.BaseEventLoop, output: asyncio.Queue):
+        while True:
+            try:
+                for trading_pair in self._trading_pairs:
+                    snapshot_msg: OrderBookMessage = await self._order_book_snapshot(trading_pair)
+                    output.put_nowait(snapshot_msg)
+                    self.logger().debug(f"Saved order book snapshot for {trading_pair}")
+                delta = CONSTANTS.ONE_HOUR - time.time() % CONSTANTS.ONE_HOUR
+                await self._sleep(delta)
+            except asyncio.CancelledError:
+                raise
+            except Exception:
+                self.logger().error(
+                    "Unexpected error occurred fetching orderbook snapshots. Retrying in 5 seconds...", exc_info=True
+                )
+                await self._sleep(5.0)
```

Comment:
Phemex sends an initial snapshot and then sends the diffs based on that initial snapshot. Lately, if an exchange offers those guarantees, we `pass` on this method since the periodic snapshots won't be necessary. WhiteBit is an example of an exchange that implements that approach.

Reply:
Snapshot are pushed every 1 minute. How do you propose we handle that?

Reply:
Understood. My mistake really, the github documentation repo I was following said it was only published once at the start of the stream. The code makes sense in the 1-minute scenario.